Fix big insertions in a single transaction

emile-tawfik opened the pull request:
Fix https://github.com/dgraph-io/dgraph/issues/4391

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/dgraph/4392)
<!-- Reviewable:end -->

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dgraph-io/dgraph?pullRequest=4392) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dgraph-io/dgraph?pullRequest=4392) before we can accept your contribution.<br/><hr/>**Emile Tawfik** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dgraph-io/dgraph?pullRequest=4392) it.</sub>

emile-tawfik commented:
Thank you for your review.
I changed name of types and variable following your directions expected for the type `indexedPlist` I decided to make it `indexedPostingList` to avoid conflict with variable that will be named `indexPlist`.

I joined a test file that is a simplification of my case.
I joined the output of the test on my machine with and without the PR. (I removed the last sub-test for the current version: it take more than 10min so it timeout)

There is the commands that I used to run my dgraph locally:
```sh
dgraph zero --bindall=true --my=127.0.0.1:5080 -w=wz
dgraph alpha --bindall=true --lru_mb 4096 --zero=127.0.0.1:5080 --my=127.0.0.1:7080 --badger.tables mmap
```

[testWithoutPR.txt](https://github.com/dgraph-io/dgraph/files/3950393/testWithoutPR.txt)
[testWithPR.txt](https://github.com/dgraph-io/dgraph/files/3950394/testWithPR.txt)
```go
package dgraphtest

import (
	"context"
	"encoding/json"
	"fmt"
	"testing"
	"time"

	"github.com/dgraph-io/dgo/v2"
	"github.com/dgraph-io/dgo/v2/protos/api"
	"google.golang.org/grpc"
	"google.golang.org/grpc/balancer/roundrobin"
)

const schema = `
type Data {
	data.value: string
}
<data.value>: string .

type DataRef {
	dataref.hash: string
	dataref.data: Data
}
<dataref.hash>: string @index(hash) .
<dataref.data>: uid .

type Node {
	node.hash: string
	node.children: [Node]
	node.datarefs: [DataRef]
}
<node.hash>: [string] @index(hash) .
<node.children>: [uid] .
<node.datarefs>: [uid] .

type TreeRef {
	treeref.allhashes: [string]
	treeref.topnode: Node
}
<treeref.allhashes>: [string] @index(hash) .
<treeref.topnode>: uid .
`

type Data struct {
	UID   string `json:"uid"`
	Value string `json:"data.value"`
}

type DataRef struct {
	UID  string `json:"uid"`
	Hash string `json:"dataref.hash"`
	Data *Data  `json:"dataref.data"`
}

type Node struct {
	UID      string     `json:"uid"`
	Hash     string     `json:"node.hash"`
	Children []*Node    `json:"node.children"`
	DataRefs []*DataRef `json:"node.datarefs"`
}

type TreeRef struct {
	UID       string   `json:"uid"`
	AllHashes []string `json:"treeref.allhashes"`
	TopNode   *Node    `json:"treeref.topnode"`
}

type treeGenerationData struct {
	nbChildrens []int
	nbDataRefs  int
}

func (tgd *treeGenerationData) String() string {
	return fmt.Sprintf("{Childrens:%+v, DataRefs:%d}", tgd.nbChildrens, tgd.nbDataRefs)
}

func createNode(testnb int, treeRef *TreeRef, dataRefs []*DataRef, path string, childrens []int) *Node {
	v := fmt.Sprintf("t%d_node%s", testnb, path)
	treeRef.AllHashes = append(treeRef.AllHashes, v)
	ret := Node{
		UID:      "_:" + v,
		Hash:     v,
		DataRefs: dataRefs,
	}
	if len(childrens) > 0 {
		for i := 0; i < childrens[0]; i++ {
			path += fmt.Sprintf("_%d", i)
			ret.Children = append(ret.Children, createNode(testnb, treeRef, dataRefs, path, childrens[1:]))
		}
	}
	return &ret
}

func (tgd *treeGenerationData) CreateTreeRef(testnb int) *TreeRef {
	dataRefs := make([]*DataRef, 0, tgd.nbDataRefs)
	for i := 0; i < tgd.nbDataRefs; i++ {
		dref := fmt.Sprintf("t%d_dataref%d", testnb, i)
		d := fmt.Sprintf("t%d_data%d", testnb, i)
		dataRefs = append(dataRefs, &DataRef{
			UID:  "_:" + dref,
			Hash: dref,
			Data: &Data{
				UID:   "_:" + d,
				Value: d,
			},
		})
	}
	treeRef := TreeRef{
		UID: "_:root",
	}
	treeRef.TopNode = createNode(testnb, &treeRef, dataRefs, "", tgd.nbChildrens)
	return &treeRef
}

const dgraphHost = `127.0.0.1:9080`

func TestDgraphBigInsertion(t *testing.T) {
	tests := []treeGenerationData{
		// Single Node
		treeGenerationData{nbChildrens: []int{1}, nbDataRefs: 0},
		treeGenerationData{nbChildrens: []int{1}, nbDataRefs: 1},
		treeGenerationData{nbChildrens: []int{1}, nbDataRefs: 10},
		treeGenerationData{nbChildrens: []int{1}, nbDataRefs: 50},
		treeGenerationData{nbChildrens: []int{1}, nbDataRefs: 100},
		treeGenerationData{nbChildrens: []int{1}, nbDataRefs: 500},
		// 5 Nodes
		treeGenerationData{nbChildrens: []int{1, 4}, nbDataRefs: 0},
		treeGenerationData{nbChildrens: []int{1, 4}, nbDataRefs: 1},
		treeGenerationData{nbChildrens: []int{1, 4}, nbDataRefs: 10},
		treeGenerationData{nbChildrens: []int{1, 4}, nbDataRefs: 50},
		treeGenerationData{nbChildrens: []int{1, 4}, nbDataRefs: 100},
		treeGenerationData{nbChildrens: []int{1, 4}, nbDataRefs: 500},
		// 10 Nodes
		treeGenerationData{nbChildrens: []int{1, 3, 2}, nbDataRefs: 0},
		treeGenerationData{nbChildrens: []int{1, 3, 2}, nbDataRefs: 1},
		treeGenerationData{nbChildrens: []int{1, 3, 2}, nbDataRefs: 10},
		treeGenerationData{nbChildrens: []int{1, 3, 2}, nbDataRefs: 50},
		treeGenerationData{nbChildrens: []int{1, 3, 2}, nbDataRefs: 100},
		treeGenerationData{nbChildrens: []int{1, 3, 2}, nbDataRefs: 500},
		// 56 Nodes
		treeGenerationData{nbChildrens: []int{1, 5, 10}, nbDataRefs: 0},
		treeGenerationData{nbChildrens: []int{1, 5, 10}, nbDataRefs: 1},
		treeGenerationData{nbChildrens: []int{1, 5, 10}, nbDataRefs: 10},
		treeGenerationData{nbChildrens: []int{1, 5, 10}, nbDataRefs: 50},
		treeGenerationData{nbChildrens: []int{1, 5, 10}, nbDataRefs: 100},
		treeGenerationData{nbChildrens: []int{1, 5, 10}, nbDataRefs: 500},
		// 511 Nodes
		treeGenerationData{nbChildrens: []int{1, 5, 100}, nbDataRefs: 0},
		treeGenerationData{nbChildrens: []int{1, 5, 100}, nbDataRefs: 1},
		treeGenerationData{nbChildrens: []int{1, 5, 100}, nbDataRefs: 10},
		treeGenerationData{nbChildrens: []int{1, 5, 100}, nbDataRefs: 50},
		treeGenerationData{nbChildrens: []int{1, 5, 100}, nbDataRefs: 100},
		treeGenerationData{nbChildrens: []int{1, 5, 100}, nbDataRefs: 500},
		// 1011 Nodes
		treeGenerationData{nbChildrens: []int{1, 10, 100}, nbDataRefs: 0},
		treeGenerationData{nbChildrens: []int{1, 10, 100}, nbDataRefs: 1},
		treeGenerationData{nbChildrens: []int{1, 10, 100}, nbDataRefs: 10},
		treeGenerationData{nbChildrens: []int{1, 10, 100}, nbDataRefs: 50},
		treeGenerationData{nbChildrens: []int{1, 10, 100}, nbDataRefs: 100},
		treeGenerationData{nbChildrens: []int{1, 10, 100}, nbDataRefs: 500},
	}

	conn, err := grpc.Dial("dns:///"+dgraphHost,
		grpc.WithBalancerName(roundrobin.Name),
		grpc.WithInsecure(),
		grpc.WithDefaultCallOptions(
			grpc.MaxCallRecvMsgSize(64<<(10*2)), // ~64Mb
			grpc.MaxCallSendMsgSize(64<<(10*2)), // ~64Mb
		),
	)
	if err != nil {
		t.Fatal(err)
	}
	dc := api.NewDgraphClient(conn)
	dg := dgo.NewDgraphClient(dc)
	ctx := context.TODO()
	if err := dg.Alter(ctx, &api.Operation{DropAll: true}); err != nil {
		t.Fatal(err)
	}
	if err := dg.Alter(ctx, &api.Operation{Schema: schema}); err != nil {
		t.Fatal(err)
	}
	for testnb, test := range tests {
		t.Run(test.String(), func(t *testing.T) {
			treeRef := test.CreateTreeRef(testnb)
			jsonTreeRef, err := json.Marshal(treeRef)
			if err != nil {
				t.Fatal(err)
			}
			txn := dg.NewTxn()
			resp, err := txn.Mutate(ctx, &api.Mutation{
				SetJson:   jsonTreeRef,
				CommitNow: true,
			})
			if err != nil {
				t.Fatal(err)
			}
			parsing := time.Duration(resp.Latency.ParsingNs)
			processing := time.Duration(resp.Latency.ProcessingNs)
			encoding := time.Duration(resp.Latency.EncodingNs)
			t.Logf("Parsing %v\tProcessing %v\tEncoding %v", parsing, processing, encoding)
		})
	}
}
```

martinmr commented:
It looks like some of our tests are failing. I'll try to take a look at it today.

emile-tawfik commented:
Do you have any news on your side ?
What kind of tests are failing ?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/dgraph-io/dgraph?pullRequest=4392) <br/>All committers have signed the CLA.

MichelDiz commented:
Hey guys, do we have some news about this PR?

emile-tawfik commented:
Sorry, I closed the issue but forgot the MR.

It's seems that the issue was fixed in the latest versions.

Thank you.
